emacs/config.org
#+TITLE:Cameron's Emacs Config
#+AUTHOR: Cameron
#+DESCRIPTION: My personal config (non-Doom)
#+STARTUP: showeverything
#+OPTIONS: toc:2

* TABLE OF CONTENTS :toc:
- [[#init][INIT]]
  - [[#source-externals][Source Externals]]
  - [[#sane-defaults][Sane Defaults]]
- [[#evil-mode][EVIL MODE]]
- [[#keybindings][KEYBINDINGS]]
- [[#theme][THEME]]
  - [[#fonts][Fonts]]
  - [[#icons][Icons]]
  - [[#colors][Colors]]
  - [[#highlight-todo][Highlight Todo]]
  - [[#modeline][Modeline]]
  - [[#visual-tweaks][Visual Tweaks]]
- [[#dashboard][DASHBOARD]]
  - [[#projectile-config][Projectile Config]]
  - [[#dashboard-1][Dashboard]]
- [[#elfeed][ELFEED]]
- [[#files][FILES]]
  - [[#backup][Backup]]
  - [[#dired][Dired]]
  - [[#neotree][NeoTree]]
- [[#ivy-counsel][IVY (COUNSEL)]]
- [[#language-support][LANGUAGE SUPPORT]]
  - [[#enable-languages][Enable Languages]]
  - [[#flycheck][Flycheck]]
  - [[#company][Company]]
- [[#org-mode][ORG MODE]]
  - [[#org-tasks][Org Tasks]]
  - [[#enable-toc][Enable ToC]]
  - [[#enable-org-bullets][Enable Org Bullets]]
  - [[#disable-electric-indent][Disable Electric Indent]]
  - [[#org-level-headers][Org Level Headers]]
  - [[#enable-org-tempo][Enable Org Tempo]]
- [[#perspective][PERSPECTIVE]]
- [[#pomodoro][POMODORO]]
- [[#shells-and-terminals][SHELLS AND TERMINALS]]
  - [[#eshell][Eshell]]
  - [[#vterm][Vterm]]
  - [[#vterm-toggle][Vterm-Toggle]]
  - [[#ghostterm][Ghostterm]]
- [[#sudo-edit][SUDO EDIT]]
- [[#tldr][TLDR]]
- [[#which-key][WHICH KEY]]

* INIT
** Source Externals
#+begin_src emacs-lisp
  (add-to-list 'load-path "~/.config/emacs/scripts/")

  (require 'elpaca-setup)
  (require 'buffer-move-setup)

#+end_src


** Sane Defaults
#+begin_src emacs-lisp
  (delete-selection-mode 1)
  (electric-indent-mode -1)
  (electric-pair-mode 0)
  (global-auto-revert-mode t)
#+end_src

* EVIL MODE

#+begin_src emacs-lisp

    ;; Install a package via the elpaca macro
  ;; See the "recipes" section of the manual for more details.

  ;; (elpaca example-package)

  ;; Install use-package support
  (elpaca elpaca-use-package
    ;; Enable use-package :ensure support for Elpaca.
    (elpaca-use-package-mode))

  ;;When installing a package used in the init file itself,
  ;;e.g. a package which adds a use-package key word,
  ;;use the :wait recipe keyword to block until that package is installed/configured.
  ;;For example:
  ;;(use-package general :ensure (:wait t) :demand t)

  ;; Expands to: (elpaca evil (use-package evil :demand t))
  (use-package evil :ensure t :demand t
    :init
    (setq evil-want-integration t)
    (setq evil-want-keybinding nil)
    (setq evil-vsplit-window-right t)
    (setq evil-split-window-below t)
    (setq evil-want-C-u-scroll t)
    (evil-mode))
  (use-package evil-collection :ensure t :demand t
    :after evil
    :config
    (setq evil-collection-mode-list `(dasboard dired ibuffer))
    (evil-collection-init))
  (use-package evil-tutor :ensure t :demand t)

  (use-package evil-commentary :ensure t :demand t
    :after evil
    :config (evil-commentary-mode))

  (with-eval-after-load 'evil-maps
    (define-key evil-motion-state-map (kbd "SPC") nil)
    (define-key evil-motion-state-map (kbd "RET") nil)
    (define-key evil-motion-state-map (kbd "TAB") nil))
  (setq org-return-follows-link t)
    

  ;;Turns off elpaca-use-package-mode current declaration
  ;;Note this will cause evaluate the declaration immediately. It is not deferred.
  ;;Useful for configuring built-in emacs features.
  (use-package emacs :ensure nil :config (setq ring-bell-function #'ignore))

#+end_src


* KEYBINDINGS

#+begin_src emacs-lisp

  (use-package general :ensure t :demand t
    :config
    (general-evil-setup)

    ;; setup 'SPC' as the global leader key
    (general-create-definer cmcn/leader-keys
      :states '(normal insert visual emacs)
      :keymaps 'override
      :prefix "SPC" ;;set leader
      :global-prefix "M-SPC") ;; access leader in insert mode

    (cmcn/leader-keys
      "SPC" '(counsel-M-x :wk "Counsel M-x")
      "." '(find-file :wk "Find file")
      "f c" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Edit emacs config")
      "f r" '(counsel-recentf :wk "Find recent files")
      )

    (cmcn/leader-keys
      "b" '(:ignore t :wk "buffer")
      "b b" '(switch-to-buffer :wk "Switch buffer")
      "b s" '(ibuffer :wk "Ibuffer")
      "b k" '(kill-current-buffer :wk "Kill this buffer")
      "b n" '(next-buffer :wk "Next buffer")
      "b p" '(previous-buffer :wk "Previous buffer")
      "b r" '(revert-buffer :wk "Reload buffer"))
    
    (cmcn/leader-keys
      "d" '(:ignore t :wk "Dired")
      "d d" '(dired :wk "Open dired")
      "d j" '(dired-jump :wk "Dired jump to current")
      "d n" '(neotree-dir :wk "Open directory in neotree")
      "d p" '(peep-dired :wk "Peep-dired")
      )

    (cmcn/leader-keys
      "e" '(:ignore t :wk "EShell/Evaluate")
      "e b" '(eval-buffer :wk "Evaluate elisp in buffer")
      "e d" '(eval-defun :wk "Evaluate defun containing or after point")
      "e e" '(eval-expression :wk "Evaluate an elisp expression")
      "e h" '(counsel-esh-history :wk "Eshell history")
      "e l" '(eval-last-sexp :wk "Evaluate elisp expression before point")
      "e r" '(eval-region :wk "Evaluate elisp in region")
      "e s" '(eshell :wk "Eshell"))
      

    (cmcn/leader-keys
      "h" '(:ignore t :wk "Help")
      "h f" '(describe-function :wk "Describe function")
      "h v" '(describe-variable :wk "Describe variable")
      "h r r" '((lambda () (interactive) (load-file "~/.config/emacs/init.el")) :wk "Reload emacs config"))
    
    (cmcn/leader-keys
      "m" '(:ignore t :wk "Org")
      "m a" '(org-agenda :wk "Org agenda")
      "m e" '(org-export-dispatch :wk "Org export dispatch")
      "m i" '(org-toggle-item :wk "Org toggle item")
      "m t" '(org-todo :wk "Org todo")
      "m B" '(org-babel-tangle :wk "Org babel tangle")
      "m T" '(org-todo-list :wk "Org todo list")
      "m c" '(org-capture :wk "Org Capture")
      "m p" '(org-priority :wk "Org Choose Priority")
      )
    
    (cmcn/leader-keys
      "m b" '(:ignore t :wk "Tables")
      "m b -" '(org-table-insert-hline :wk "Insert hline in table")
    )

    (cmcn/leader-keys
      "m d" '(:ignore t :wk "Date/deadline")
      "m d t" '(org-time-stamp :wk "Org time stamp")
    )

    (cmcn/leader-keys
      "t" '(:ignore t :wk "Toggle")
      "t l" '(display-line-numbers-mode :wk "Toggle line numbers")
      "t t" '(visual-line-mode :wk "Toggle truncated lines")
      "t v" '(vterm-toggle :wk "Toggle vterm")
      "t n" '(neotree-toggle :wk "Toggle neotree")
      "t p" '(org-pomodoro :wk "Toggle pomodoro")
      )
    
    (cmcn/leader-keys
      "w" '(:ignore t :wk "Windows")
      "w c" '(evil-window-delete :wk "Close window")
      "w n" '(evil-window-new :wk "New window")
      "w v" '(evil-window-split :wk "Vertical split window")
      "w s" '(evil-window-vsplit :wk "Horizontal split window")

      "w h" '(evil-window-left :wk "Window left")
      "w j" '(evil-window-down :wk "Window down")
      "w k" '(evil-window-up :wk "Window up")
      "w l" '(evil-window-right :wk "Window right")
      "w w" '(evil-window-next :wk "Goto next window")

      "w H" '(buf-move-left :wk "Buffer move left")
      "w J" '(buf-move-down :wk "Buffer move down")
      "w K" '(buf-move-up :wk "Buffer move up")
      "w L" '(buf-move-right :wk "Buffer move right"))
  )

  ;; Stop needing to hit escape 3 times to exit mini buffer
  (global-set-key [escape] 'keyboard-escape-quit)
#+end_src


* THEME
Defining the various fonts Emacs will use.

** Fonts

#+begin_src emacs-lisp
  (set-face-attribute 'default nil
    :font "FiraMono Nerd Font"
    :height 180
    :weight 'medium)
  (set-face-attribute 'variable-pitch nil
    :font "Helvetica"
    :height 180
    :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
    :font "FiraMono Nerd Font"
    :height 180
    :weight 'medium)

  (set-face-attribute 'font-lock-comment-face nil
  		    :slant 'italic)
  (set-face-attribute 'font-lock-keyword-face nil
  		    :slant 'italic)

  (add-to-list 'default-frame-alist '(font . "FiraMono Nerd Font 18"))

  (setq-default line-spacing 0.12)

  (global-set-key (kbd "C-=") 'text-scale-increase)
  (global-set-key (kbd "C--") 'text-scale-decrease)
  (global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
  (global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)


#+end_src

** Icons

Be sure to run 'nerd-icons-install-fonts' if icons are missing

#+begin_src emacs-lisp
  (use-package nerd-icons :demand t :ensure t
    :if (display-graphic-p))

  (use-package nerd-icons-dired :demand t :ensure t
    :hook (dired-mode . nerd-icons-dired-mode))
#+end_src

** Colors

#+begin_src emacs-lisp
  (use-package catppuccin-theme :demand t :ensure t
    :config
    (catppuccin-load-flavor 'mocha)
  )

  (use-package doom-themes :demand t :ensure t
    :config
    (setq doom-themes-enable-bold t
  	doom-themes-enable-italic t)
    ;; (load-theme 'doom-one t)
    (doom-themes-neotree-config)
    (doom-themes-org-config))

  (use-package rainbow-mode :demand t :ensure t
    :diminish
    :hook org-mode prog-mode)

  (use-package rainbow-delimiters :demand t :ensure t
    :hook ((emacs-lisp-mode . rainbow-delimiters-mode)
  	 (clojure-mode . rainbow-delimiters-mode)))

#+end_src

** Highlight Todo
#+begin_src emacs-lisp
  (use-package hl-todo :demand t :ensure t
    :hook ((org-mode . hl-todo-mode)
  	 (prog-mode . hl-todo-mode))
    :config
    (setq hl-todo-highlight-punctuation ":"
  	hl-todo-keyword-faces
  	`(("TODO" warning bold)
  	  ("FIXME" error bold)
  	  ("HACK" font-lock-constant-face bold)
  	  ("REVIEW" font-lock-keyword-face bold)
  	  ("NOTE" success bold)
  	  ("DEPRECATED" font-lock-doc-face bold))))

#+end_src

** Modeline

#+begin_src emacs-lisp
  (use-package doom-modeline :demand t :ensure t
    :init (doom-modeline-mode 1)
    :after nerd-icons
    :config
    (setq doom-modeline-height 30
  	doom-modeline-bar-width 5
  	doom-modeline-persp-name t
  	doom-modeline-persp-icon t))

#+end_src


** Visual Tweaks

#+begin_src emacs-lisp
  (add-to-list 'default-frame-alist '(alpha-background . 90))


  (menu-bar-mode -1)
  (tool-bar-mode -1)
  (scroll-bar-mode -1)

  (setq display-line-numbers-type 'relative)
  (global-display-line-numbers-mode 1)
  (global-visual-line-mode t)

  (use-package diminish :demand t :ensure t)
#+end_src

* DASHBOARD

** Projectile Config
#+begin_src emacs-lisp
  (use-package projectile :demand t :ensure t
    :diminish
    :config
    (projectile-mode 1))

#+end_src

** Dashboard
#+begin_src emacs-lisp
  (use-package dashboard :demand t :ensure t
    :after nerd-icons
    :init
    (setq initial-buffer-choice 'dashboard-open)
    (setq dashboard-icon-type 'nerd-icons)
    (setq dashboard-set-heading-icons t)
    (setq dashboard-set-file-icons t)
    (setq dashboard-banner-logo-title "Always be learning!")
    (setq dashboard-startup-banner 'logo)
    ;; (setq dashboard-startup-banner "file.png")
    (setq dashboard-center-content t)
    (setq dashboard-items '((recents . 5)
  			  (agenda . 5)
  			  (bookmarks . 3)
  			  (projects . 3)
  			  (registers . 3)))
    :config
    ;; (dashboard-modify-heading-icons '((recents . "nf-fa-clock")
    ;; 				    (bookmarks . "nf-fa-book")))
    (dashboard-setup-startup-hook))

#+end_src

* ELFEED
#+begin_src emacs-lisp
    (use-package elfeed :demand t :ensure t
      :config
      (setq elfeed-search-feed-face ":foreground #ffffff :weight bold"
    	elfeed-feeds (quote
    		      (
  		       ("https://www.opensource.com/feed" opensource linux)
    		      ))))
#+end_src

* FILES
** Backup
#+begin_src emacs-lisp
  (setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))
#+end_src

** Dired

#+begin_src emacs-lisp
  (use-package dired-open :demand t :ensure t
    :after nerd-icons
    :config
    (setq dired-open-extensions '(("gif" . "sxiv")
  				("jpg" . "sxiv")
  				("png" . "sxiv")
  				("mkv" . "mpv")
  				("mp4" . "mpv"))))

  (use-package peep-dired :demand t :ensure t
    :after dired
    :hook (evil-normalize-keymaps . peep-dired-hook)
    :config
    (evil-define-key 'normal dired-mode-map (kbd "h") 'dired-up-directory)
    (evil-define-key 'normal dired-mode-map (kbd "l") 'dired-open-file)
    (evil-define-key 'normal peep-dired-mode-map (kbd "j") 'peep-dired-next-file)
    (evil-define-key 'normal peep-dired-mode-map (kbd "k") 'peep-dired-prev-file))

#+end_src

** NeoTree

#+begin_src emacs-lisp
  (use-package neotree :demand t :ensure t
    :config
    (setq neo-smart-open t
  	neo-show-hidden-files t
  	neo-window-width 55
  	neo-window-fixed-size nil
  	inhibit-compacting-font-caches t
  	projectile-switch-project-action 'neotree-projectile-action)
    (add-hook 'neo-after-create-hook
  	    #'(lambda (_)
  		(with-current-buffer (get-buffer neo-buffer-name)
  		  (setq truncate-lines t)
  		  (setq word-wrap nil)
  		  (make-local-variable 'auto-hscroll-mode)
  		  (setq auto-hscroll-mode nil)))))

#+end_src
* IVY (COUNSEL)
+ Ivy, a generic completion mechanism for Emacs.
+ Counsel, a collection of Ivy-enhanced versions of common Emacs commands.
+ Ivy-rich allows us to add descriptions alongside the commands in M-x.

#+begin_src emacs-lisp
  (use-package counsel :demand t :ensure t
    :diminish
    :after ivy
    :config (counsel-mode))

  (use-package ivy :demand t :ensure t
    :diminish
    :bind
    (("C-c C-r" . ivy-resume)
     ("C-x B" . ivy-switch-buffer-other-window))
    :custom
    (setq ivy-use-virtual-buffers t)
    (setq ivy-count-format "(%d/%d) ")
    (setq enable-recursive-minibuffers t)
    :config
    (ivy-mode))

  (use-package nerd-icons-ivy-rich :demand t :ensure t
    :init (nerd-icons-ivy-rich-mode 1))


emacs/config.el
(add-to-list 'load-path "~/.config/emacs/scripts/")

(require 'elpaca-setup)
(require 'buffer-move-setup)

(delete-selection-mode 1)
(electric-indent-mode -1)
(electric-pair-mode 0)
(global-auto-revert-mode t)

;; Install a package via the elpaca macro
;; See the "recipes" section of the manual for more details.

;; (elpaca example-package)

;; Install use-package support
(elpaca elpaca-use-package
  ;; Enable use-package :ensure support for Elpaca.
  (elpaca-use-package-mode))

;;When installing a package used in the init file itself,
;;e.g. a package which adds a use-package key word,
;;use the :wait recipe keyword to block until that package is installed/configured.
;;For example:
;;(use-package general :ensure (:wait t) :demand t)

;; Expands to: (elpaca evil (use-package evil :demand t))
(use-package evil :ensure t :demand t
  :init
  (setq evil-want-integration t)
  (setq evil-want-keybinding nil)
  (setq evil-vsplit-window-right t)
  (setq evil-split-window-below t)
  (setq evil-want-C-u-scroll t)
  (evil-mode))
(use-package evil-collection :ensure t :demand t
  :after evil
  :config
  (setq evil-collection-mode-list `(dasboard dired ibuffer))
  (evil-collection-init))
(use-package evil-tutor :ensure t :demand t)

(use-package evil-commentary :ensure t :demand t
  :after evil
  :config (evil-commentary-mode))

(with-eval-after-load 'evil-maps
  (define-key evil-motion-state-map (kbd "SPC") nil)
  (define-key evil-motion-state-map (kbd "RET") nil)
  (define-key evil-motion-state-map (kbd "TAB") nil))
(setq org-return-follows-link t)
  

;;Turns off elpaca-use-package-mode current declaration
;;Note this will cause evaluate the declaration immediately. It is not deferred.
;;Useful for configuring built-in emacs features.
(use-package emacs :ensure nil :config (setq ring-bell-function #'ignore))

(use-package general :ensure t :demand t
  :config
  (general-evil-setup)

  ;; setup 'SPC' as the global leader key
  (general-create-definer cmcn/leader-keys
    :states '(normal insert visual emacs)
    :keymaps 'override
    :prefix "SPC" ;;set leader
    :global-prefix "M-SPC") ;; access leader in insert mode

  (cmcn/leader-keys
    "SPC" '(counsel-M-x :wk "Counsel M-x")
    "." '(find-file :wk "Find file")
    "f c" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Edit emacs config")
    "f r" '(counsel-recentf :wk "Find recent files")
    )

  (cmcn/leader-keys
    "b" '(:ignore t :wk "buffer")
    "b b" '(switch-to-buffer :wk "Switch buffer")
    "b s" '(ibuffer :wk "Ibuffer")
    "b k" '(kill-current-buffer :wk "Kill this buffer")
    "b n" '(next-buffer :wk "Next buffer")
    "b p" '(previous-buffer :wk "Previous buffer")
    "b r" '(revert-buffer :wk "Reload buffer"))
  
  (cmcn/leader-keys
    "d" '(:ignore t :wk "Dired")
    "d d" '(dired :wk "Open dired")
    "d j" '(dired-jump :wk "Dired jump to current")
    "d n" '(neotree-dir :wk "Open directory in neotree")
    "d p" '(peep-dired :wk "Peep-dired")
    )

  (cmcn/leader-keys
    "e" '(:ignore t :wk "EShell/Evaluate")
    "e b" '(eval-buffer :wk "Evaluate elisp in buffer")
    "e d" '(eval-defun :wk "Evaluate defun containing or after point")
    "e e" '(eval-expression :wk "Evaluate an elisp expression")
    "e h" '(counsel-esh-history :wk "Eshell history")
    "e l" '(eval-last-sexp :wk "Evaluate elisp expression before point")
    "e r" '(eval-region :wk "Evaluate elisp in region")
    "e s" '(eshell :wk "Eshell"))
    

  (cmcn/leader-keys
    "h" '(:ignore t :wk "Help")
    "h f" '(describe-function :wk "Describe function")
    "h v" '(describe-variable :wk "Describe variable")
    "h r r" '((lambda () (interactive) (load-file "~/.config/emacs/init.el")) :wk "Reload emacs config"))
  
  (cmcn/leader-keys
    "m" '(:ignore t :wk "Org")
    "m a" '(org-agenda :wk "Org agenda")
    "m e" '(org-export-dispatch :wk "Org export dispatch")
    "m i" '(org-toggle-item :wk "Org toggle item")
    "m t" '(org-todo :wk "Org todo")
    "m B" '(org-babel-tangle :wk "Org babel tangle")
    "m T" '(org-todo-list :wk "Org todo list")
    "m c" '(org-capture :wk "Org Capture")
    "m p" '(org-priority :wk "Org Choose Priority")
    )
  
  (cmcn/leader-keys
    "m b" '(:ignore t :wk "Tables")
    "m b -" '(org-table-insert-hline :wk "Insert hline in table")
  )

  (cmcn/leader-keys
    "m d" '(:ignore t :wk "Date/deadline")
    "m d t" '(org-time-stamp :wk "Org time stamp")
  )

  (cmcn/leader-keys
    "t" '(:ignore t :wk "Toggle")
    "t l" '(display-line-numbers-mode :wk "Toggle line numbers")
    "t t" '(visual-line-mode :wk "Toggle truncated lines")
    "t v" '(vterm-toggle :wk "Toggle vterm")
    "t n" '(neotree-toggle :wk "Toggle neotree")
    "t p" '(org-pomodoro :wk "Toggle pomodoro")
    )
  
  (cmcn/leader-keys
    "w" '(:ignore t :wk "Windows")
    "w c" '(evil-window-delete :wk "Close window")
    "w n" '(evil-window-new :wk "New window")
    "w v" '(evil-window-split :wk "Vertical split window")
    "w s" '(evil-window-vsplit :wk "Horizontal split window")

    "w h" '(evil-window-left :wk "Window left")
    "w j" '(evil-window-down :wk "Window down")
    "w k" '(evil-window-up :wk "Window up")
    "w l" '(evil-window-right :wk "Window right")
    "w w" '(evil-window-next :wk "Goto next window")

    "w H" '(buf-move-left :wk "Buffer move left")
    "w J" '(buf-move-down :wk "Buffer move down")
    "w K" '(buf-move-up :wk "Buffer move up")
    "w L" '(buf-move-right :wk "Buffer move right"))
)

;; Stop needing to hit escape 3 times to exit mini buffer
(global-set-key [escape] 'keyboard-escape-quit)

(set-face-attribute 'default nil
  :font "FiraMono Nerd Font"
  :height 180
  :weight 'medium)
(set-face-attribute 'variable-pitch nil
  :font "Helvetica"
  :height 180
  :weight 'medium)
(set-face-attribute 'fixed-pitch nil
  :font "FiraMono Nerd Font"
  :height 180
  :weight 'medium)

(set-face-attribute 'font-lock-comment-face nil
		    :slant 'italic)
(set-face-attribute 'font-lock-keyword-face nil
		    :slant 'italic)

(add-to-list 'default-frame-alist '(font . "FiraMono Nerd Font 18"))

(setq-default line-spacing 0.12)

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
(global-set-key (kbd "<C-wheel-up>") 'text-scale-increase)
(global-set-key (kbd "<C-wheel-down>") 'text-scale-decrease)

(use-package nerd-icons :demand t :ensure t
  :if (display-graphic-p))

(use-package nerd-icons-dired :demand t :ensure t
  :hook (dired-mode . nerd-icons-dired-mode))

(use-package catppuccin-theme :demand t :ensure t
  :config
  (catppuccin-load-flavor 'mocha)
)

(use-package doom-themes :demand t :ensure t
  :config
  (setq doom-themes-enable-bold t
	doom-themes-enable-italic t)
  ;; (load-theme 'doom-one t)
  (doom-themes-neotree-config)
  (doom-themes-org-config))

(use-package rainbow-mode :demand t :ensure t
  :diminish
  :hook org-mode prog-mode)

(use-package rainbow-delimiters :demand t :ensure t
  :hook ((emacs-lisp-mode . rainbow-delimiters-mode)
	 (clojure-mode . rainbow-delimiters-mode)))

(use-package hl-todo :demand t :ensure t
  :hook ((org-mode . hl-todo-mode)
	 (prog-mode . hl-todo-mode))
  :config
  (setq hl-todo-highlight-punctuation ":"
	hl-todo-keyword-faces
	`(("TODO" warning bold)
	  ("FIXME" error bold)
	  ("HACK" font-lock-constant-face bold)
	  ("REVIEW" font-lock-keyword-face bold)
	  ("NOTE" success bold)
	  ("DEPRECATED" font-lock-doc-face bold))))

(use-package doom-modeline :demand t :ensure t
  :init (doom-modeline-mode 1)
  :after nerd-icons
  :config
  (setq doom-modeline-height 30
	doom-modeline-bar-width 5
	doom-modeline-persp-name t
	doom-modeline-persp-icon t))

(add-to-list 'default-frame-alist '(alpha-background . 90))


(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(setq display-line-numbers-type 'relative)
(global-display-line-numbers-mode 1)
(global-visual-line-mode t)

(use-package diminish :demand t :ensure t)

(use-package projectile :demand t :ensure t
  :diminish
  :config
  (projectile-mode 1))

(use-package dashboard :demand t :ensure t
  :after nerd-icons
  :init
  (setq initial-buffer-choice 'dashboard-open)
  (setq dashboard-icon-type 'nerd-icons)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Always be learning!")
  (setq dashboard-startup-banner 'logo)
  ;; (setq dashboard-startup-banner "file.png")
  (setq dashboard-center-content t)
  (setq dashboard-items '((recents . 5)
			  (agenda . 5)
			  (bookmarks . 3)
			  (projects . 3)
			  (registers . 3)))
  :config
  ;; (dashboard-modify-heading-icons '((recents . "nf-fa-clock")
  ;; 				    (bookmarks . "nf-fa-book")))
  (dashboard-setup-startup-hook))

(use-package elfeed :demand t :ensure t
  :config
  (setq elfeed-search-feed-face ":foreground #ffffff :weight bold"
	elfeed-feeds (quote
		      (
		       ("https://www.opensource.com/feed" opensource linux)
		      ))))

(setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))

(use-package dired-open :demand t :ensure t
  :after nerd-icons
  :config
  (setq dired-open-extensions '(("gif" . "sxiv")
				("jpg" . "sxiv")
				("png" . "sxiv")
				("mkv" . "mpv")
				("mp4" . "mpv"))))

(use-package peep-dired :demand t :ensure t
  :after dired
  :hook (evil-normalize-keymaps . peep-dired-hook)
  :config
  (evil-define-key 'normal dired-mode-map (kbd "h") 'dired-up-directory)
  (evil-define-key 'normal dired-mode-map (kbd "l") 'dired-open-file)
  (evil-define-key 'normal peep-dired-mode-map (kbd "j") 'peep-dired-next-file)
  (evil-define-key 'normal peep-dired-mode-map (kbd "k") 'peep-dired-prev-file))

(use-package neotree :demand t :ensure t
  :config
  (setq neo-smart-open t
	neo-show-hidden-files t
	neo-window-width 55
	neo-window-fixed-size nil
	inhibit-compacting-font-caches t
	projectile-switch-project-action 'neotree-projectile-action)
  (add-hook 'neo-after-create-hook
	    #'(lambda (_)
		(with-current-buffer (get-buffer neo-buffer-name)
		  (setq truncate-lines t)
		  (setq word-wrap nil)
		  (make-local-variable 'auto-hscroll-mode)
		  (setq auto-hscroll-mode nil)))))

(use-package counsel :demand t :ensure t
  :diminish
  :after ivy
  :config (counsel-mode))

(use-package ivy :demand t :ensure t
  :diminish
  :bind
  (("C-c C-r" . ivy-resume)
   ("C-x B" . ivy-switch-buffer-other-window))
  :custom
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)
  :config
  (ivy-mode))

(use-package nerd-icons-ivy-rich :demand t :ensure t
  :init (nerd-icons-ivy-rich-mode 1))

(use-package ivy-rich :demand t :ensure t
  :after ivy
  :init (ivy-rich-mode 1)
  :custom
  (ivy-virtual-abbreviate 'full
   ivy-rich-switch-buffer-align-virtual-buffer t
   ivy-rich-path-style 'abbrev)
  :config
  (ivy-set-display-transformer 'ivy-switch-buffer
			       'ivy-rich-switch-buffer-transformer))

(use-package lua-mode :demand t :ensure t)
(use-package rust-mode :demand t :ensure t)

(use-package flycheck :demand t :ensure t
  :defer t
  :diminish
  :init (global-flycheck-mode))

(use-package company :demand t :ensure t
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t))

(use-package company-box :demand t :ensure t
  :after company
  :diminish
  :hook (company-mode . company-box-mode))

(setq org-directory "~/Documents/99_meta/org/")

(setq org-agenda-files
      (list (concat org-directory "tasks.org")
	    (concat org-directory "notes.org")
	    (concat org-directory "inbox.org")
	    (concat org-directory "journal.org")))

(setq org-default-notes-file (concat org-directory "inbox.org"))

(setq org-capture-templates
      `(
	("i" "idea" entry
	 (file ,org-default-notes-file)
	 "* %? :idea:\n%U\n")

	("j" "journal" entry
	 (file+olp+datetree ,(concat org-directory "journal.org"))
	 "* %U\n%?\n")

	("n" "idea" entry
	 (file ,org-default-notes-file)
	 "* %? :note:\n%U\n%a\n")

	("t" "task" entry
	 (file+headline ,(concat org-directory "tasks.org") "Inbox")
	 "* TODO %?\n%^t\n%a\n")
	))

(use-package toc-org :demand t :ensure t
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets :demand t :ensure t :diminish)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
(eval-after-load 'org-indent '(diminish 'org-indent-mode))

(electric-indent-mode -1)

(custom-set-faces
 '(org-level-1 ((t (:inherit outline-1 :height 1.7))))
 '(org-level-2 ((t (:inherit outline-2 :height 1.6))))
 '(org-level-3 ((t (:inherit outline-3 :height 1.5))))
 '(org-level-4 ((t (:inherit outline-4 :height 1.4))))
 '(org-level-5 ((t (:inherit outline-5 :height 1.3))))
 '(org-level-6 ((t (:inherit outline-5 :height 1.2))))
 '(org-level-7 ((t (:inherit outline-5 :height 1.1)))))

(require 'org-tempo)

(use-package perspective :demand t :ensure t
  :custom
  (persp-mode-prefix-key (kbd "C-c M-p"))
  :init
  (persp-mode)
  :config
  (setq persp-state-default-file "~/.config/emacs/sessions"))
(add-hook 'ibuffer-hook
	  (lambda ()
	    (persp-ibuffer-set-filter-groups)
	    (unless (eq ibuffer-sorting-mode 'alphabetic)
	      (ibuffer-do-sort-by-alphabetic))))

(use-package org-pomodoro :demand t :ensure t
  :config
  (setq org-pomodoro-length 25
	org-pomodoro-short-break-length 5
	org-pomodoro-long-break-length 15

	;; org-pomodoro-audio-player "mpv"
	
	;; org-pomodoro-finished-sound "~/"
	;; org-pomodoro-short-break-sound "~/"
	;; org-pomodoro-long-break-sound "~/"
	
	org-pomodoro-finished-sound-p t
	org-pomodoro-short-break-sound-p t
	org-pomodoro-long-break-sound-p t))

(use-package eshell-syntax-highlighting :demand t :ensure t
  :after esh-mode
  :config
  (eshell-syntax-highlighting-global-mode +1))

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))

(use-package vterm :demand t :ensure t
  :config
(setq shell-file-name "/bin/zsh"
      vterm-max-scrollback 5000))

(use-package vterm-toggle :demand t :ensure t
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p nil)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
	       '((lambda (buffer-or-name _)
		   (let ((buffer (get-buffer buffer-or-name)))
		     (with-current-buffer buffer
		       (or (equal major-mode 'vterm-mode)
			   (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
		 (display-buffer-reuse-window display-buffer-at-bottom)
		 (reusable-frames . visible)
		 (window-height . 0.3))))

(use-package sudo-edit :demand t :ensure t
  :config
  (cmcn/leader-keys
    "f u" '(sudo-edit-find-file :wk "Sudo find file")
    "f U" '(sudo-edit :wk "Sudo edit file")))

(use-package tldr :demand t :ensure t)

(use-package which-key :demand t :ensure t
  :diminish
  :init
  (which-key-mode 1)

  :config
  (setq which-key-side-window-location 'bottom
	whichkey-sort-order #'which-key-key-order-alpha
	which-key-sort-uppercase-first nil
	which-key-add-column-padding 1
	which-key-max-display-columns nil
	which-key-min-display-lines 6
	which-key-side-window-slot -10
	which-key-side-window-max-height 0.25
	which-key-idle-delay 0.5
	which-key-max-description-length 25
	which-key-allow-imprecise-window-fit nil
	which-key-separator " -> " ))
